architecture/spmc.rst

```diff
@@ -0,0 +1,269 @@
+	SPs.
+- Memory management:
+	The SPMC is responsible for the memory management of the SPs. Memory
+	can be shared between SPs and between a SP to the normal world.
+
+This document describes OP-TEE as a S-EL1 SPMC.
+
+Secure Partitions
+=================
+Secure Partitions (SPs) are the endpoints used in the FF-A protocol. When
+OP-TEE is used as a SPMC SPs run primarily inside S-EL0.
+
+The SPMC also exposes the OP-TEE core, privileged mode, as an secure endpoint
+itself. This is used to handle all GlobalPlaform programming mode operations.
+All GlobalPlatform messages are encapsulated inside FF-A messages.
+The OP-TEE endpoint will unpack the messages and afterwards handle them as
+standard OP-TEE calls. This is needed as TF-A (S-EL3) does only allow
+FF-A messages to be passed to the secure world when the SPMD is enabled.
+
+SPs run from the initial boot of the system until power down and don't have any
+build-in session management compared to GPD TEE TAs. The only means of
+communicating with the outside world is through the FF-A spec defined messages.
+SPs context is saved between executions.
+
+The
+`Trusted Service <https://www.trustedfirmware.org/projects/trusted-services/>`_
+repository includes the libSP libary which export all needed functions to build
+a S-EL0 SP, it also includes many examples of how to implement SPs.
+
+SPMC Program Flow
+=================
+SP images are stored OP-TEE image as early TAs are: the binary images are
+embedded in OP-TEE Core a read-only data section.
+This makes it possible to load SPs during boot. Early TSs are embedded inside
+the OP-TEE image. This makes it possible to load SPs during the initialisation
+of OP-TEE and no rich-os is needed in the normal world. Ldelf is used to load
+and initialise the SPs.
+
+Starting SPs
+------------
+SPs are loaded and started as the last step in OP-TEE's initialisation process.
+This is done by adding ``sp_init_all()`` to the ``boot_final`` initcall level.
+
+.. uml::
+        :width: 800
+
+        autoactivate on
+        thread_optee_smc_a64.s -> boot_final
+        boot_final -> secure_partition.c:			sp_init_all()
+        loop for each SP
+        secure_partition.c -> secure_partition.c:		sp_init_uuid()
+        return
+        end
+        return
+        return
+        autoactivate off
+        thread_optee_smc_a64.s -> thread_optee_smc_a64.s:	thread_ffa_msg_wait()
+        thread_optee_smc_a64.s -> thread_optee_smc_a64.s:	ffa_msg_loop()
+        autoactivate off
+        thread_optee_smc_a64.s -> SPMD:				SMC
+
+:``sp_init_all()``:		Initialise all SPs which have been added by the
+				``SP_PATHS`` compiler option and run them
+:``thread_ffa_msg_wait()``:	All SPs are loaded and started. A
+				``FFA_MSG_WAIT`` message is sent to the Normal
+				World.
+
+
+Each SP is loaded into the system using ldelf and started. This is done using
+Ldelf and is based around the same process as loading the early TAs.
```

Comment:
Mentioning ldelf here again is a bit repetitive. You may want to rephrase these two sentences.

Reply:
Done